libs/server/Resp/AsyncProcessor.cs

```diff
@@ -0,0 +1,137 @@
+        /// Whether async mode is turned on for the session
+        /// </summary>
+        bool useAsync = false;
+
+        /// <summary>
+        /// How many async operations are started and completed
+        /// </summary>
+        long asyncStarted = 0, asyncCompleted = 0;
+
+        /// <summary>
+        /// Async waiter for async operations
+        /// </summary>
+        SingleWaiterAutoResetEvent asyncWaiter = null;
+
+        /// <summary>
+        /// Semaphore for barrier command to wait for async operations to complete
+        /// </summary>
+        SemaphoreSlim asyncDone = null;
+
+
+        /// <summary>
+        /// Handle a async network GET command that goes pending
+        /// </summary>
+        /// <typeparam name="TGarnetApi"></typeparam>
+        /// <param name="storageApi"></param>
+        void NetworkGETPending<TGarnetApi>(ref TGarnetApi storageApi)
+            where TGarnetApi : IGarnetApi
+        {
+            unsafe
+            {
+                while (!RespWriteUtils.WriteError($"ASYNC {asyncStarted++}", ref dcurr, dend))
+                    SendAndReset();
+            }
+            if (asyncStarted == 1) // first async operation on the session, create the IO continuation processor
+            {
+                asyncWaiter = new()
+                {
+                    RunContinuationsAsynchronously = true
+                };
+                var _storageApi = storageApi;
+                _ = Task.Run(async () => await AsyncGetProcessor(_storageApi));
+            }
+            else
+            {
+                Debug.Assert(asyncWaiter != null);
+                asyncWaiter.Signal();
+            }
+        }
+
+        /// <summary>
+        /// Background processor for async IO continuations. This is created only when async is turned on for the session.
+        /// It handles all the IO completions and takes over the network sender to send async responses when ready.
+        /// Note that async responses are not guaranteed to be in the same order that they are issued.
+        /// </summary>
+        async Task AsyncGetProcessor<TGarnetApi>(TGarnetApi storageApi)
+            where TGarnetApi : IGarnetApi
+        {
+            while (true)
+            {
+                while (asyncCompleted < asyncStarted)
+                {
+                    // First complete all pending ops
+                    storageApi.GET_CompletePending(out var completedOutputs, true);
+
+                    try
+                    {
+                        unsafe
+                        {
+                            // We are ready to send responses so we take over the network sender from the main ProcessMessage thread
+                            // Note that we cannot take over while ProcessMessage is in progress because partial responses may have been
+                            // sent at that point. For sync responses that span multiple ProcessMessage calls (e.g., MGET), we need the
```

Comment:
It's not clear how that is enforced; in fact this thread may be running before the NetworkGET call that launched it has returned from ProcessMessages

Reply:
It is enforced using the SpinLock inside NetworkSender. The network processing thread is holding the lock, so the async processor cannot actually do anything until it gets that lock. right?